src/merge.rs

```diff
@@ -192,3 +194,295 @@ impl<'repo> Drop for AnnotatedCommit<'repo> {
+    pub fn mode(&mut self, mode: Option<FileMode>) -> &mut MergeFileInput {
+        self.mode = mode;
+
+        if let Some(mode) = self.mode {
+            self.raw.mode = mode.into();
+        }
+
+        self
+    }
+
+    /// File content, text or binary
+    pub fn content(&mut self, content: Option<Vec<u8>>) -> &mut MergeFileInput {
+        self.content = content;
+
+        self.raw.size = self.content.as_ref().map(|c| c.len()).unwrap_or(0);
+        self.raw.ptr = self
+            .content
+            .as_ref()
+            .map(|c| c.as_ptr() as *const _)
+            .unwrap_or(ptr::null());
+
+        self
+    }
+
+    /// Get the raw struct in C
+    pub fn raw(&self) -> *const raw::git_merge_file_input {
+        &self.raw as *const _
+    }
+}
+
+impl MergeFileInput {
+    /// Create from Repository and IndexEntry
+    pub fn from(repo: &Repository, index_entry: &IndexEntry) -> MergeFileInput {
+        let blob = repo
+            .find_blob(index_entry.id.clone())
+            .expect("failed to find blob of index entry to make MergeFileInput");
+        let content = blob.content().to_vec();
+
+        let mut input = MergeFileInput::new();
+        input.content(Some(content));
+        input.path(index_entry.path.clone());
+        input.mode(Some(index_entry.mode.into()));
+
+        input
+    }
+}
+
+impl std::fmt::Debug for MergeFileInput {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let mut ds = f.debug_struct("MergeFileInput");
+        if let Some(path) = &self.path {
+            ds.field("path", path);
+        }
+        ds.field("mode", &self.mode);
+
+        if let Some(mode) = self.mode {
+            match mode {
+                FileMode::Unreadable => {}
+                FileMode::Tree => {}
+                FileMode::Blob => {
+                    let content = self
+                        .content
+                        .as_ref()
+                        .map(|s| String::from_utf8_lossy(&s).to_string())
+                        .unwrap_or("unknown content".to_string());
+                    ds.field("content", &content);
+                }
+                FileMode::BlobExecutable => {}
+                FileMode::Link => {}
+                FileMode::Commit => {}
+            }
+        }
+        ds.finish()
+    }
+}
+
+/// For git_merge_file_result
+pub struct MergeFileResult {
+    /// True if the output was automerged, false if the output contains
+    /// conflict markers.
+    pub automergeable: bool,
+
+    /// The path that the resultant merge file should use, or NULL if a
+    /// filename conflict would occur.
+    pub path: Option<String>,
+
+    /// The mode that the resultant merge file should use.
+    pub mode: FileMode,
+
+    /// The contents of the merge.
+    pub content: Option<Vec<u8>>,
+}
+
+impl MergeFileResult {
+    /// Create MergeFileResult from C
+    pub fn from_raw(raw: raw::git_merge_file_result) -> MergeFileResult {
```

Comment:
I think this'll want to be `unsafe`?

Reply:
I have used `unsafe` at the specified lines, is that ok?
```rust
impl MergeFileResult {
    /// Create MergeFileResult from C
    pub fn from_raw(raw: raw::git_merge_file_result) -> MergeFileResult {
        let c_str: &CStr = unsafe { CStr::from_ptr(raw.path) };  // unsafe
        let str_slice: &str = c_str.to_str().unwrap_or("unknown path");
        let path: String = str_slice.to_owned();

        let content =
            unsafe { slice::from_raw_parts(raw.ptr as *const u8, raw.len as usize).to_vec() }; // unsafe

        MergeFileResult {
            automergeable: raw.automergeable > 0,
            path: Some(path),
            mode: raw.mode.into(),
            content: Some(content),
        }
    }
}
```

Reply:
Ah the `unsafe` part is that this is taking a raw C structure where the pointers aren't guaranteeed to be valid, so the `unsafe` signals "the caller needs to take care of that validity"

Reply:
Understood, than you!